AI/VCAI/VCAI.cpp

```diff
@@ -781,129 +781,212 @@ void VCAI::makeTurn()
-	for(const CGTownInstance * t : cb->getTownsInfo())
-		moveCreaturesToHero(t);
 
-	try
+	//TODO: move these checks to mainLoop()
+
+	auto reservedHeroesCopy = reservedHeroesMap; //work on copy => the map may be changed while iterating (eg because hero died when attempting a goal)
+	for(auto hero : reservedHeroesCopy)
 	{
-		/*Below loop causes heroes with objects locked to them to keep trying to realize previous goal. By design when object is locked another heroes do not attempt to visit it.
-		Object lock happens on turn when some hero gets assigned visit tile with appropiate map object. So basically all heroes that had VisitTile goal with object assigned and not completed
-		will be in the loop. Sometimes heroes get assigned more important objectives, but still keep reserved objects for later. There is a problem with that - they have
-		reserved objects in the list, so they fall to this loop at start of turn and visiting object isn't delayed. Comments for that function are supposed to help newer VCAI maintainers*/
+		if(reservedHeroesMap.count(hero.first))
+			continue; //hero might have been removed while we were in this loop
+		if(!hero.first.validAndSet())
+		{
+			logAi->error("Hero %s present on reserved map. Shouldn't be.", hero.first.name);
+			continue;
+		}
 
-		//Pick objects reserved in previous turn - we expect only nearby objects there
-		auto reservedHeroesCopy = reservedHeroesMap; //work on copy => the map may be changed while iterating (eg because hero died when attempting a goal)
-		for(auto hero : reservedHeroesCopy)
+		std::vector<const CGObjectInstance *> vec(hero.second.begin(), hero.second.end());
+		boost::sort(vec, CDistanceSorter(hero.first.get()));
+		for(auto obj : vec)
 		{
-			if(reservedHeroesMap.count(hero.first))
-				continue; //hero might have been removed while we were in this loop
-			if(!hero.first.validAndSet())
+			if(!obj || !cb->getObj(obj->id))
 			{
-				logAi->error("Hero %s present on reserved map. Shouldn't be.", hero.first.name);
+				logAi->error("Error: there is wrong object on list for hero %s", hero.first->name);
 				continue;
 			}
-
-			std::vector<const CGObjectInstance *> vec(hero.second.begin(), hero.second.end());
-			boost::sort(vec, CDistanceSorter(hero.first.get()));
-			for(auto obj : vec)
-			{
-				if(!obj || !cb->getObj(obj->id))
-				{
-					logAi->error("Error: there is wrong object on list for hero %s", hero.first->name);
-					continue;
-				}
-				striveToGoal(sptr(Goals::VisitTile(obj->visitablePos()).sethero(hero.first)));
-			}
+			striveToGoal(sptr(Goals::VisitTile(obj->visitablePos()).sethero(hero.first)));
 		}
+	}
+}
 
-		//now try to win
-		/*below line performs goal decomposition, result of the function is ONE goal for ONE hero to realize.*/
-		striveToGoal(sptr(Goals::Win()));
+void VCAI::mainLoop()
+{
+	std::vector<Goals::TSubgoal> basicGoals, elementarGoals; //no duplicates allowed (operator ==)

```

Comment:
Why not use std::set then?

Reply:
std::set needs < operator... which I just implemented afterwards :P

Reply:
Ahh yes, now I remember. I want to remove duplicates using == operator, which is stronger (removes also non-identical goals).